### ISSUE TITLE
SUBST: a notion of identity

### ISSUE THREAD

**Issue by SwampertX** [source-academy/js-slang#340]
Take a look at the following [program](https://sourceacademy.nus.edu.sg/playground#chap=1&ext=NONE&prgrm=GYVwdgxgLglg9mABMAFMMBKRBvATgUyhFyXQG4BfAKFWCwF5Hk0Myg):

```javascript
function f(fn) { return fn; }
f(f) === f(f); // true
```
returns `false` in the substituter.

**Comment by SwampertX** [comment]
This might be a feature, not a bug.

The substituter uses lambda-calculus style substitution, there is no sense of environment. Equivalence of functions? We might be talking about [alpha-equivalence](https://en.wikipedia.org/wiki/Lambda_calculus#Alpha_equivalence) here.

Below is what "equivalence" in the substitution model might have meant in the textbook:

The above program returns true in node implementation (and native implementation) because the name f, points to the specific block in memory that stores the function (or equivalently, the function object saved in the environment). The term "environment" has been mentioned even in the [first chapter of the textbook](https://sicp.comp.nus.edu.sg/chapters/4).

basically: the substituter has no sense of environment, so function equality by environment (memory location or similar) is impossible. However alpha equivalence (correct me if I'm wrong) fails when we have recursive definitions. for example:

```javascript
function f() {return g();}
function g() {return f();}
f === g; // ?
```
This makes it impossible to determine using alpha equivalence, imo.

**Comment by martin-henz** [comment]
We need to get to the bottom of this.

In our framework, don't we have the following property?

When we evaluate
lhs === rhs
and both lhs and rhs are syntax graphs that represent functions (definitions or declarations), then the two syntax graphs are the same graphs (pointer equality) if and only if both sides originate from evaluating the same definition/declaration?

Can we check pointer equality of the syntax graphs to find out whether two functions are identical?

**Comment by martin-henz** [comment]
The bigger picture: 

Is the substituter able to represent a notion of identity that is consistent with the environment model? How can we explain the notion of "substitution" if it preserves identity?

**Comment by martin-henz** [comment]
Handled by https://github.com/source-academy/js-slang/issues/389
